Sources/X509/Verifier/VerificationDiagnostic.swift

```diff
@@ -0,0 +1,237 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import SwiftASN1
+
+public struct VerificationDiagnostic: Sendable {
+    struct LeafCertificateHasUnhandledCriticalExtensions: Hashable {
+        var leafCertificate: Certificate
+        var handledCriticalExtensions: [ASN1ObjectIdentifier]
+    }
+    struct LeafCertificateIsInTheRootStoreButDoesNotMeetPolicy: Hashable {
+        var leafCertificate: Certificate
+        var failsToMeetPolicyReason: String
+    }
+    struct ChainFailsToMeetPolicy: Hashable {
+        var chain: UnverifiedCertificateChain
+        var failsToMeetPolicyReason: String
+    }
+    struct IssuerHasUnhandledCriticalExtension: Hashable {
+        var issuer: Certificate
+        var partialChain: [Certificate]
+        var handledCriticalExtensions: [ASN1ObjectIdentifier]
+    }
+    struct IssuerHasNotSignedCertificate: Hashable {
+        var issuer: Certificate
+        var partialChain: [Certificate]
+    }
+    enum Storage: Hashable {
+        case leafCertificateHasUnhandledCriticalExtension(LeafCertificateHasUnhandledCriticalExtensions)
+        case leafCertificateIsInTheRootStoreButDoesNotMeetPolicy(LeafCertificateIsInTheRootStoreButDoesNotMeetPolicy)
+        case chainFailsToMeetPolicy(ChainFailsToMeetPolicy)
+        case intermediateHashUnhandledCriticalExtension(IssuerHasUnhandledCriticalExtension)
+        case issuerHasNotSignedCertificate(IssuerHasNotSignedCertificate)
+    }
+    var storage: Storage
+}
+
+extension VerificationDiagnostic {
+    static func leafCertificateHasUnhandledCriticalExtension(
+        _ leafCertificate: Certificate,
+        handledCriticalExtensions: [ASN1ObjectIdentifier]
```

Comment:
As a diagnostic for unhandled extensions I would expect to see whatever isn't handled but this is `handled`... is that a typo?

Reply:
Not a typo but on purpose. We delay the computation of the unhandled extensions until `description` is accessed to make constructing a `LeafCertificateHasUnhandledCriticalExtension` just a couple of retains. `handledCriticalExtensions` and `leaf` are already constructed when we produce the diagnostics. Unhandled extensions are computed in the `Certificate.Extensions.unhandledCriticalExtensions(for:)` at the very bottom of this file.